src/lib.rs

```diff
@@ -409,8 +463,19 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
             error!("saving render context failed: {:?}", e);
             return;
         }
-        paint_ctx.transform(Affine::translate(self.state.layout_rect.origin().to_vec2()));
-        self.paint(paint_ctx, data, env);
+        let layout_origin = self.state.layout_rect.origin().to_vec2();
+        paint_ctx.transform(Affine::translate(layout_origin));
+
+        let visible = paint_ctx
+            .region()
+            .to_rect()
+            .intersect(self.state.layout_rect)
```

Comment:
Hmm. First, we talked about not always culling to the layout rect, as this shouldn't be a hard invariant. Second, if we are intersecting, then this might be a good place to stop the recursion (rather than having the caller check).

Basically what I think we should do here is have two variants, one which does intersect layout rect (and culls recursion), one that just applies `layout_origin`.

Reply:
good point, hadn't thought of that. will take another look.

Reply:
So thinking about this more, I think the intersection was just a mistake. My intention here was to always make the caller check, if they're interested. I can see the merit of doing the check in the paint call itself, though, I'll experiment with that (as a separate fn)

Reply:
@raphlinus I've updated this to add `paint_with_offset_always`; `paint_with_offset` gets the short-circuiting behaviour. This should mean that most widgets will 'just work', not getting painted if they aren't visible.

I'm not sure what to call the new function. 🤷‍♂

Reply:
Heh, I think it's ok if it has a somewhat awkward name. We're going to need to rework that section more than a bit when we do fine-grained invalidation.

I'm ok with this going in at this point.

Reply:
I'm surprised you aren't a bit more circumspect about my context closure business.

I would have preferred to have some method on `PaintCtx` that took the new region and returned a new `PaintCtx` but I couldn't figure that out.